### PR TITLE
Make ERC2771Context return original sender address if `msg.data.length <= 20`

### PR DESCRIPTION
**Pull request by ernestognw**
Fixes an audit finding. Report will be published soon.

The current implementation of the ERC2771Forwarder doesn't check that the `msg.data` has a length of at least the ERC2771 suffixed address, nullifying the result and producing the `address(0)` whenever accessing an out-of-bounds `calldata` location (i.e. calldata with size 19).

This PR checks for the correct length and falls back to the original sender.

### CODE REVIEW

File: .changeset/unlucky-beans-obey.md
Comment (Amxx) [inline .changeset/unlucky-beans-obey.md:5]: I'm not sure what "`msg.data.length` is not appended with" means. The address is appended to te msg.data, and the msg.data.length increases by 20.

```suggestion
`ERC2771Context`: Return the forwarder address whenever the `msg.data` or calls sent by a trusted forwarder is not long enough to contain the request signer address (i.e. `msg.data.length` is less than 20 bytes).
```

File: test/metatx/ERC2771Context.test.js
Comment (Amxx) [inline test/metatx/ERC2771Context.test.js:91]: In that case the recipient is called directly, so you can do
```suggestion
        const { receipt } = await recipient.msgSender({ from: anotherAccount });

        await expectEvent(receipt, 'Sender', { sender: anotherAccount });
```
(which is faster)

**Review by Amxx — DISMISSED** [review]

**Comment by ernestognw** [comment]
I thought accepting a suggestion doesn't dismiss reviews, but it does.
Can you reapprove @Amxx?

File: .changeset/unlucky-beans-obey.md
Comment (frangio) [inline .changeset/unlucky-beans-obey.md]: The current description is a bit misleading because it sounds like we can know whether the request signer address is appended, but the only thing we can know is if the length is greater than 20 bytes.

```suggestion
`ERC2771Context`: Return the forwarder address whenever `msg.data` is smaller than 20 bytes as specified by ERC-2771, since the request signer address can't be extracted.
```

Comment (frangio) [reply .changeset/unlucky-beans-obey.md]: Reviewing the new changeset...

Comment (ernestognw) [reply .changeset/unlucky-beans-obey.md]: Note that the suggestion you're making is missing that the change only applies when called via the forwarder. Now that I think about it, it seems that the forwarder address is always returned if the calldata length is less than 20 bytes.

File: .changeset/unlucky-beans-obey.md
Comment (frangio) [inline .changeset/unlucky-beans-obey.md:5]:
```suggestion
`ERC2771Context`: Return the forwarder address whenever the `msg.data` of a call originating from a trusted forwarder is not long enough to contain the request signer address (i.e. `msg.data.length` is less than 20 bytes), as specified by ERC-2771.
```

Comment (ernestognw) [reply .changeset/unlucky-beans-obey.md:5]: Thoughts on?

```suggestion
`ERC2771Context`: Return the forwarder address whenever the `msg.data` of a call originated from a trusted forwarder is not long enough to contain the request signer address (i.e. `msg.data.length` is less than 20 bytes), as specified by ERC-2771
```

Comment (Amxx) [reply .changeset/unlucky-beans-obey.md:5]: Either is good for me. When you reach concensus you can merge without my formal approval.

Comment (ernestognw) [reply .changeset/unlucky-beans-obey.md]: No issues 👌

**Review by frangio — APPROVED** [review]

**Review by Amxx — APPROVED** [review]

**Comment by frangio** [comment]
For future observers, the title of this PR says `<= 20` but it should have been `< 20`. The code does the latter (correctly).

**Pull request merged**
